compiler/rustc_mir_transform/src/single_enum.rs

```diff
@@ -0,0 +1,44 @@
+use rustc_middle::mir::*;
+use rustc_middle::ty::TyCtxt;
+use rustc_mir_dataflow::impls::SingleEnumVariant;
+use rustc_mir_dataflow::{lattice::FlatSet, Analysis};
+
+use crate::MirPass;
+
+pub struct SingleEnum;
+
+impl<'tcx> MirPass<'tcx> for SingleEnum {
+    fn run_pass(&self, tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
+        let mut single_enum_variants = SingleEnumVariant::new(tcx, body)
+            .into_engine(tcx, body)
+            .iterate_to_fixpoint()
+            .into_results_cursor(body);
+
+        let to_replace = body.basic_blocks().iter_enumerated().filter_map(|(bb, block)| {
+            let TerminatorKind::SwitchInt { discr, targets, .. } = &block.terminator().kind else { return None };
```

Comment:
Seems like this could be made more general by replacing `discriminant(x)` expressions with the known discriminant, instead of looking specifically for `switchInt`. That would also allow optimizing, e.g., `x == Foo::A`, in addition to `match`es.

Reply:
Are switch statements combined with Discriminant followed by SwitchInt? I can add that in addition, or should I replace it completely?

Reply:
See https://godbolt.org/z/x3E9qK4xf -- `match` of an enum becomes:
```
_2 = discriminant(_1);
switchInt(move _2) -> [0_isize: bb3, 1_isize: bb4, 2_isize: bb1, otherwise: bb2];
```

If your pass replaces that with:
```
_2 = const 1_isize;
switchInt(move _2) -> [0_isize: bb3, 1_isize: bb4, 2_isize: bb1, otherwise: bb2];
```
then another pass (I believe `SimplifyConstCondition`) can optimize that to goto `bb4` directly, as in https://godbolt.org/z/8fa15ef1o

I think it's generally preferred to make each pass do the minimal transform necessary, so I would make this pass just replace `Discriminant`, and leave the further `SwitchInt` simplification to `SimplifyConstCondition`, if that's what you're asking.

Reply:
sg, I guess my immediate thought was that matches are just done using just `SwitchInts`, so I can do this.